modules/sdk-core/src/bitgo/walletUtil/utxoMessageProof.ts

```diff
@@ -0,0 +1,102 @@
+import * as utxolib from '@bitgo/utxo-lib';
+
+import { IWallet } from '../wallet/iWallet';
+import { BitGoBase } from '../bitgoBase';
+
+type MessageInfo = {
+  message: string;
+  address: string;
+};
+
+type Claim = {
+  originWalletId: string;
+  status: string;
+  originAddress?: string;
+  allocationAmount: string;
+};
+
+export interface IMessageProvider {
+  /**
+   * Returns the messages and addresses that we want to sign. We call this function multiple times until there are no more
+   * messages. If there are no more messages, an empty array is returned. Note that we only return messages in sets of 200.
+   */
+  getMessagesAndAddressesToSign(): Promise<MessageInfo[]>;
+}
+
+export class MidnightMessageProvider implements IMessageProvider {
+  protected messageCache: MessageInfo[];
+  protected network: utxolib.Network;
+  protected midnightClaimUrl: string;
+  protected prevId: string | undefined;
+  constructor(private bitgo: BitGoBase, private coinName: string, private walletId: string, private message: string) {
+    this.messageCache = [];
+    this.network = utxolib.networks[coinName];
+    this.midnightClaimUrl = `https://app.bitgo${
+      utxolib.isTestnet(this.network) ? '-test' : ''
+    }.com/api/airdrop-claim/v1/midnight/claims/${coinName}/${walletId}`;
+  }
+
+  async getMessagesAndAddressesToSign(): Promise<MessageInfo[]> {
+    if (this.messageCache.length > 0) {
+      // Get the first 200 messages in this array. Remove them from the cache.
+      return this.messageCache.splice(0, 200);
+    }
+    const query: Record<string, unknown> = { status: 'UNINITIALIZED', limit: 1000 };
+    if (this.prevId) {
+      query.prevId = this.prevId;
+    }
+    const response = await this.bitgo.get(this.midnightClaimUrl).query(query).result();
+    if (response.status !== 'success') {
+      throw new Error(`Unexpected status code ${response.status} from ${this.midnightClaimUrl}`);
+    }
+    if (response?.pagination?.hasNext) {
+      this.prevId = response?.pagination?.nextPrevId;
+    }
+
+    this.messageCache = response.claims.map((claim: Claim) => {
+      if (!claim.originAddress) {
+        throw new Error(`Claim ${JSON.stringify(claim)} is missing originAddress`);
+      }
+      return {
+        message: this.message,
+        address: claim.originAddress,
+      };
+    });
+    return this.messageCache.splice(0, 200);
+  }
+}
+
+export async function bulkSignBip322MidnightMessages(
+  wallet: IWallet,
+  message: string,
+  walletPassphrase?: string
+): Promise<{ success: boolean; numMessagesSigned: number }> {
+  const provider = new MidnightMessageProvider(wallet.bitgo, wallet.coin(), wallet.id(), message);
+  return bulkSignBip322MessagesWithProvider(provider, wallet);
+}
+
+export async function bulkSignBip322MessagesWithProvider(
```

Comment:
This function should be agnostic of where the payload is coming from. It should just take that as an input and generate the transaction for signing.